modules/workergroup/nodepools.tf

```diff
@@ -0,0 +1,153 @@
+
+    dynamic "placement_configs" {
+      iterator = ad_iterator
+      for_each = [for n in lookup(each.value, "placement_ads", local.ad_numbers) :
+        length(local.ad_numbers) == 1 ? local.ad_number_to_name[1] : local.ad_number_to_name[n]
+      ]
+      content {
+        availability_domain = ad_iterator.value
+        subnet_id           = var.primary_subnet_id
+      }
+    }
+
+    nsg_ids = lookup(each.value, "worker_nsg_ids", var.worker_nsg_ids)
+
+    # flannel requires cni type only
+    dynamic "node_pool_pod_network_option_details" {
+      for_each = coalesce(var.cni_type, "flannel") == "flannel" ? [1] : []
+      content {
+        cni_type = "FLANNEL_OVERLAY"
+      }
+    }
+
+    # native requires max pods/node, nsg ids, subnet ids
+    dynamic "node_pool_pod_network_option_details" {
+      for_each = coalesce(var.cni_type, "flannel") == "npn" ? [1] : []
+      content {
+        cni_type          = "OCI_VCN_IP_NATIVE"
+        max_pods_per_node = min(max(var.max_pods_per_node, 1), 110)
+        # pick the 1st pod nsg here until https://github.com/oracle/terraform-provider-oci/issues/1662 is clarified and resolved
+        pod_nsg_ids = slice(
+          lookup(each.value, "pod_nsg_ids", var.pod_nsg_ids),
+          0, min(1, length(lookup(each.value, "pod_nsg_ids", var.pod_nsg_ids)))
+        )
+        pod_subnet_ids = tolist([coalesce(lookup(each.value, "pod_subnet_id", var.pod_subnet_id), var.primary_subnet_id)])
+      }
+    }
+
+    freeform_tags = merge(coalesce(local.freeform_tags, {}), lookup(each.value, "freeform_tags", { oke_group = each.key }))
+    defined_tags  = merge(coalesce(local.defined_tags, {}), lookup(each.value, "defined_tags", {}))
+  }
+
+  # cloud-init
+  node_metadata = {
+    user_data = data.cloudinit_config.worker_np.rendered
+  }
+
+  node_shape = lookup(each.value, "shape", local.shape)
+  dynamic "node_shape_config" {
+    for_each = length(regexall("Flex", lookup(each.value, "shape", local.shape))) > 0 ? [1] : []
+    content {
+      ocpus = max(1, lookup(each.value, "ocpus", local.ocpus))
+      memory_in_gbs = ( # If > 64GB memory/core, correct input to exactly 64GB memory/core
+        (lookup(each.value, "memory", local.memory) / lookup(each.value, "ocpus", local.ocpus)) > 64
+        ? (lookup(each.value, "ocpus", local.ocpus) * 64)
+        : lookup(each.value, "memory", local.memory)
+      )
+    }
+  }
+
+  # Optimized OKE images
+  dynamic "node_source_details" {
+    for_each = var.image_type == "oke" ? [1] : []
+    content {
+      source_type             = data.oci_containerengine_node_pool_option.np_options.sources[0].source_type
+      boot_volume_size_in_gbs = lookup(each.value, "boot_volume_size", local.boot_volume_size)
+      # check for GPU,A1 and other shapes. In future, if some other shapes or images are added, we need to modify
+      image_id = ((var.image_type == "oke" && length(regexall("GPU|A1", lookup(each.value, "shape", local.shape))) == 0)
```

Comment:
Can we please request this https://github.com/oracle/terraform-provider-oci/issues/1285 is expedited? I don't fancy maintaining this if we have new options or remove existing options in the future

Reply:
Agreed... Will see what I can find.